opendatahub-io/data-science-pipelines-operator
# (feat): add tls to api server

## The issue resolved by this Pull Request:
Resolves: https://issues.redhat.com/browse/RHOAIENG-4968

Depends on: https://github.com/opendatahub-io/data-science-pipelines/pull/40



## Description of your changes:
Add new component-wide level tls enablement field to DSPA: `PodToPodTLS`, which is enabled by default and ensures that API server serves over TLS. Also updates PA and kfp UI, as well as OAUTH Server in apiserver pod to be able to recognize the CA Bundle.

One potentially controversial decision here is the choice to have the OAUTH Proxy side car communicate via svc dns name instead of localhost, the reason for this is because the certs used to serve tls for api server are leveraged from openshift service signer, and these only recognize subjects for the service name and not localhost. My assumption here is that it shouldn't matter and kube proxy / network optimizations here do not let the intra pod traffic leave the pod's worker node. 

Another thing to note, the PA deployment is updated recognize certs from 2 paths now, and because we use rhel containers, the rhel ssl default cert path is recognized here. 

Other changes are to the health check,  to allow for polling based on transfer protocol. 

## Testing instructions

<details> 

<summary>instructions.md</summary>

1. Deploy DSPO from this PR
2. Deploy DSPA using images built from here https://github.com/opendatahub-io/data-science-pipelines/pull/40/files
3. Confirm API Server pod logs show TLS serving messages: 
```
"TLS cert key/pair loaded"
...
"Starting RPC server (TLS enabled)"
"Starting Https Proxy "
```
When `podToPodTLS: false` these omit tls mentions. Note that when `podToPodTLS` is not specified in the DSPA we see https and tls grpc enabled by default. 

4. Port-forward the api server, and curl this route, confirm you see self-signed cert warnings

```
dspaNamespace=dspa # fill this out
dspaName=pr-40       # fill this out
oc port-forward -n ${dspaNamespace} service/ds-pipeline-${dspaName} 8888:8888
curl --location --request GET 'https://localhost:8888/apis/v2beta1/runs' --header "Authorization: Bearer $(oc whoami --show-token)"
```
Note the https. Attempt to use http, you should get an error response. 
```
curl --location --request GET 'http://localhost:8888/apis/v2beta1/runs' --header "Authorization: Bearer $(oc whoami --show-token)"

```
You should see "Client sent an HTTP request to an HTTPS server."
This confirms that the api http server container is now behind a tls cert. 
Try again by disabling tls verification: 

```
curl --insecure --location --request GET 'https://localhost:8888/apis/v2beta1/runs' --header "Authorization: Bearer $(oc whoami --show-token)"
```

This should return `{}` if you have no runs in this DSP.

6. Provide the signer cert pulled from the configmap `kube-root-ca.crt` to curl via `--ca-cert` and try again, attempt to list runs via the route, confirm this works:

```bash
cd $(mktemp -d)

# the following command requires yq if you don't have it, just copy this from the kube-root-ca.crt configmap in the dspa namespace
oc -n ${dspaNamespace} get configmap kube-root-ca.crt  -o yaml | yq '.data."ca.crt"' > kube-root-ca.crt 
DS_ROUTE=$(echo https://$(oc get routes -n ${dspaNamespace} ds-pipeline-${dspaName} --template={{.spec.host}}))
curl --cacert kube-root-ca.crt --location --request GET ${DS_ROUTE}/apis/v2beta1/runs --header "Authorization: Bearer $(oc whoami --show-token)"
```

7. Inspect the kfp-ui, it should have successfully connected to the server
8. Verify persistent agent works by starting a run, and confirming the status of the run is updating (e.g. green check mark) 
9. Now try deploying a dspa with `podToPodTLS=false`, confirm dsp works as before 

> Note that in the integration tests `podToPodTLS` is disabled, because we rely on the[ openshift service signing certs](https://docs.openshift.com/container-platform/4.15/security/certificates/service-serving-certificate.html) not available in a kind environment


</details>



## Checklist
